### PR TITLE
Refactor loss computation in  enhancement tasks.

### PR DESCRIPTION
**Pull request by LiChenda**
Hello!
Previously, we have implemented various models in ESPNet-SE. However, there are so many different behaviors in the loss computation stage for different models.
For example, some separation models need PIT, while others not;
Some models use time-domain loss while some use frequency-domain loss;
In frequency-domain losses, there are also many options ... 
These different behaviors were controlled by IF-ELSE conditions in the `espnet2/enh/espnet_model.py`, the amount IF-ELSE branches became too large to be maintained in a single file.

So, in this PR, I refactored the loss compuatation. The different loss functions are  abstracted into   `AbsEnhLoss` and `AbsLossWrapper` classes.
`AbsEnhLoss`  can be implemnted as a specific criterion, and `AbsLossWrapper` is a wrapper of a simple loss function to achieve more complex functions (e.g. PIT solver, or MIXIT solver in the future ) .
In `espnet2/enh/espnet_model.py`, the loss can be computed by unified interfaces.
Thus, the `espnet2/enh/espnet_model.py` file are shorten to ~200 lines.

Besides, now the loss functions can be choosen in the training config file, and flexible multi-task learning is also supported now:

```yaml


# A list for criterions
# The overlall loss in the multi-task learning will be:
# loss = weight_1 * loss_1 + ... + weight_N * loss_N
# The default `weight` for each sub-loss is 1.0
criterions: 
  # The first criterion
  - name: mse 
    conf:
      compute_on_mask: True
      mask_type: PSM
    # the wrapper for the current criterion
    # PIT is widely used in the speech separation task
    wrapper: pit
    wrapper_conf:
      weight: 1.0
  # The second criterion
  - name: l1 
    conf:
      compute_on_mask: False
    wrapper: pit
    wrapper_conf:
      weight: 1.0
      # whether the permutation is computed independently
      # If false, it will use the permutation order from 
      # the pervious criterion (if it exists)
      independent_perm: False  
  # The third criterion
  - name: si_snr 
    conf:
      eps: 1.0e-7
    wrapper: pit
    wrapper_conf:
      weight: 5.0
      independent_perm: False
```

Some existing functions are removed temporarily：

- Extra labels in neural-beamformers: dereverbration and denoising labels
- STFT-consistency for frequency-domain models

@Emrys365 and I will fix them in following PRs.

### CODE REVIEW

**Comment by sw005320** [comment]
This is really cool!
Thanks for the great PR!

File: espnet2/tasks/enh.py [hunk 1]
```diff
@@ -67,6 +74,20 @@
...
+loss_wrapper_choices = ClassChoices(
+    name="loss_wrappers",
+    classes=dict(pit=PITSolver, fixed_order=FixedOrderSolver),
+    type_check=AbsLossWrapper,
+    default=None,
+)
+
+criterion_choices = ClassChoices(
+    name="criterions",
+    classes=dict(si_snr=SISNRLoss, mse=FrequencyDomainMSE, l1=FrequencyDomainL1),
```
Comment (Emrys365) [inline espnet2/tasks/enh.py:86]: TODO: add CISDRLoss, Spectrum/Magnitude losses, TimeDomainL1

File: test/espnet2/enh/test_espnet_model.py [hunk 2]
```diff
@@ -192,23 +165,19 @@ def test_single_channel_model(
 )
 
 
+pit_wrapper = PITSolver(criterion=FrequencyDomainMSE(compute_on_mask=True))
+
+
 @pytest.mark.parametrize("training", [True, False])
 @pytest.mark.parametrize("mask_type", ["IBM", "IRM", "IAM", "PSM", "PSM^2"])
 @pytest.mark.parametrize(
     "loss_type", ["mask_mse", "magnitude", "spectrum", "spectrum_log"]
```
Comment (Emrys365) [inline test/espnet2/enh/test_espnet_model.py:174]: Not effective in current version. Different loss types are not tested.

File: test/espnet2/enh/test_espnet_model.py [hunk 3]
```diff
@@ -222,17 +191,13 @@ def test_forward_with_beamformer_net(
     inputs = random_speech[..., :ch].float()
     ilens = torch.LongTensor([16, 12])
     speech_refs = [torch.randn(2, 16, ch).float() for spk in range(num_spk)]
-    noise_ref1 = torch.randn(2, 16, ch, dtype=torch.float)
-    dereverb_ref1 = torch.randn(2, 16, ch, dtype=torch.float)
+    # noise_ref1 = torch.randn(2, 16, ch, dtype=torch.float)
+    # dereverb_ref1 = torch.randn(2, 16, ch, dtype=torch.float)
```
Comment (Emrys365) [inline test/espnet2/enh/test_espnet_model.py:195]: Better to preserve these lines for later support of denoising and dereverberation losses.

File: test/espnet2/enh/test_espnet_model.py [hunk 4]
```diff
@@ -269,7 +234,5 @@ def test_forward_with_beamformer_net(
         "speech_mix": inputs,
         "speech_mix_lengths": ilens,
         **{"speech_ref{}".format(i + 1): speech_refs[i] for i in range(num_spk)},
-        "noise_ref1": noise_ref1,
-        "dereverb_ref1": dereverb_ref1,
```
Comment (Emrys365) [inline test/espnet2/enh/test_espnet_model.py:273]: Better to preserve these lines for later support of denoising and dereverberation losses.

File: espnet2/enh/loss/wrappers/pit_solver.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                        [
+                            self.criterion(
+                                ref[s][batch].unsqueeze(0), inf[t][batch].unsqueeze(0)
+                            )
+                            for s, t in enumerate(p)
+                        ]
+                    ).mean()
+                    for batch, p in enumerate(perm)
+                ]
+            )
```
Comment (Emrys365) [inline espnet2/enh/loss/wrappers/pit_solver.py:62]: Later, we shall merge the Hungarian algorithm (https://github.com/LiChenda/espnet/blob/develop/espnet2/asr/espnet_enh_asr_model.py#L486) into this part.

File: espnet2/enh/loss/criterions/tf_domain.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        Args:
+            ref: (Batch, T, F) or (Batch, T, C, F)
+            inf: (Batch, T, F) or (Batch, T, C, F)
+        Returns:
+            loss: (Batch,)
+        """
+        assert ref.shape == inf.shape, (ref.shape, inf.shape)
+        if not is_torch_1_3_plus:
+            # in case of binary masks
+            ref = ref.type(inf.dtype)
```
Comment (Emrys365) [inline espnet2/enh/loss/criterions/tf_domain.py:179]: ditto

File: espnet2/enh/loss/criterions/tf_domain.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
```
Comment (Emrys365) [inline espnet2/enh/loss/criterions/tf_domain.py:129]: We don't need this check any more. (see https://github.com/espnet/espnet/pull/3355#discussion_r679649166)

File: espnet2/enh/espnet_model.py [hunk 8]
```diff
@@ -215,488 +132,52 @@ def forward(
...
-        # add stats for logging
-        if self.loss_type not in ("snr", "ci_sdr", "si_snr"):
-            if self.training:
-                si_snr = None
-            else:
-                speech_pre = [self.decoder(ps, speech_lengths)[0] for ps in speech_pre]
-                speech_ref = torch.unbind(speech_ref, dim=1)
-                if speech_ref[0].dim() == 3:
-                    # For si_snr loss, only select one channel as the reference
-                    speech_ref = [sr[..., self.ref_channel] for sr in speech_ref]
```
Comment (Emrys365) [inline espnet2/enh/espnet_model.py:237]: We should reuse this part because sometimes the reference can be single-channel while the input is multi-channel.

File: espnet2/enh/espnet_model.py [hunk 9]
```diff
@@ -215,488 +132,52 @@ def forward(
...
-                loss, perm = None, None
-                return loss, speech_pre, None, speech_lengths, perm
-
-            # speech_pre: list[(batch, sample)]
-            assert speech_pre[0].dim() == 2, speech_pre[0].dim()
-
-            if speech_ref.dim() == 4:
-                # For si_snr loss of multi-channel input,
-                # only select one channel as the reference
-                speech_ref = speech_ref[..., self.ref_channel]
```
Comment (Emrys365) [inline espnet2/enh/espnet_model.py:443]: We should reuse this part because sometimes the reference can be multi-channel while the network output is single-channel.

File: espnet2/enh/espnet_model.py [hunk 10]
```diff
@@ -1,42 +1,26 @@
...
-from espnet2.enh.layers.complex_utils import is_complex
+from espnet2.enh.loss.criterions.tf_domain import FrequencyDomainLoss
+from espnet2.enh.loss.criterions.time_domain import TimeDomainLoss
+from espnet2.enh.loss.wrappers.abs_wrapper import AbsLossWrapper
 from espnet2.enh.separator.abs_separator import AbsSeparator
 from espnet2.torch_utils.device_funcs import force_gatherable
 from espnet2.train.abs_espnet_model import AbsESPnetModel
 
 
 is_torch_1_3_plus = LooseVersion(torch.__version__) >= LooseVersion("1.3.0")
```
Comment (Emrys365) [inline espnet2/enh/espnet_model.py:21]: We can remove this line.

File: espnet2/tasks/enh.py (hunk 1 above)
Comment (LiChenda) [reply espnet2/tasks/enh.py]: SNRLoss, CISDRLoss added.

File: test/espnet2/enh/test_espnet_model.py (hunk 2 above)
Comment (LiChenda) [reply test/espnet2/enh/test_espnet_model.py:174]: Add unit tests for criterions and wrappers, different loss types are now tested in thier unit tests.

File: test/espnet2/enh/test_espnet_model.py (hunk 3 above)
Comment (LiChenda) [reply test/espnet2/enh/test_espnet_model.py]: Done.

File: test/espnet2/enh/test_espnet_model.py (hunk 4 above)
Comment (LiChenda) [reply test/espnet2/enh/test_espnet_model.py]: Done.

File: espnet2/enh/loss/criterions/tf_domain.py (hunk 6 above)
Comment (LiChenda) [reply espnet2/enh/loss/criterions/tf_domain.py]: Done.

File: espnet2/enh/loss/criterions/tf_domain.py (hunk 7 above)
Comment (LiChenda) [reply espnet2/enh/loss/criterions/tf_domain.py]: Done.

File: espnet2/enh/espnet_model.py (hunk 8 above)
Comment (LiChenda) [reply espnet2/enh/espnet_model.py:237]: Done.

File: espnet2/enh/espnet_model.py (hunk 9 above)
Comment (LiChenda) [reply espnet2/enh/espnet_model.py:443]: Done.

File: espnet2/enh/espnet_model.py (hunk 10 above)
Comment (LiChenda) [reply espnet2/enh/espnet_model.py]: Done.

**Review by sw005320 — APPROVED** [review]

**Comment by sw005320** [comment]
LGTM. If @Emrys365 is OK, I'll merge this PR.

**Review by Emrys365 — APPROVED** [review]
Sure, I think it is okay to merge now.

**Comment by sw005320** [comment]
OK!
Thanks!
